packages/block-editor/src/components/block-edit/index.js

```diff
@@ -20,6 +24,76 @@ import { BlockEditContextProvider, useBlockEditContext } from './context';
  */
 export { useBlockEditContext };
 
+function BlockControlFilters( props ) {
+	const { name, isSelected, clientId } = props;
+	const shouldDisplayControls = useSelect(
+		( select ) => {
+			if ( isSelected ) {
+				return true;
+			}
+
+			const {
+				getBlockName,
+				isFirstMultiSelectedBlock,
+				getMultiSelectedBlockClientIds,
+				hasSelectedInnerBlock,
+			} = select( blockEditorStore );
+
+			if ( isFirstMultiSelectedBlock( clientId ) ) {
+				return getMultiSelectedBlockClientIds().every(
+					( id ) => getBlockName( id ) === name
+				);
+			}
+
+			return (
+				hasBlockSupport(
+					getBlockName( clientId ),
+					'__experimentalExposeControlsToChildren',
+					false
+				) && hasSelectedInnerBlock( clientId )
+			);
+		},
+		[ clientId, isSelected, name ]
+	);
+
+	const hookName = 'editor.BlockControls';
+	const [ , forceRender ] = useReducer( () => [] );
+
+	useEffect( () => {
+		const namespace = 'core/block-edit/block-controls';
+
+		function onHooksUpdated( updatedHookName ) {
+			if ( updatedHookName === hookName ) {
+				forceRender();
+			}
+		}
+
+		addAction( 'hookRemoved', namespace, onHooksUpdated );
+		addAction( 'hookAdded', namespace, onHooksUpdated );
+
+		return () => {
+			removeAction( 'hookRemoved', namespace );
+			removeAction( 'hookAdded', namespace );
+		};
+	}, [] );
+
+	if ( ! shouldDisplayControls ) {
+		return;
+	}
+
+	const blockControlFilters = filters[ hookName ];
+
+	if ( ! blockControlFilters ) {
+		return;
+	}
+
+	return blockControlFilters.handlers.map(
```

Comment:
That's probably not quite how WordPress filters were designed to work. I'm not sure I've ever seen usage where the previous value would not get set at all.